staging/src/k8s.io/client-go/util/connrotation/connrotation.go

```diff
@@ -81,10 +82,27 @@ func (c *ConnectionTracker) CloseAll() {
 	c.mu.Unlock()
 
 	for conn := range conns {
-		conn.Close()
+		_ = conn.Close()
 	}
 }
 
+// CloseAllGraceful forcibly closes all tracked connections that have been marked by GetCertOrTLSHandshakeComplete.
+func (c *ConnectionTracker) CloseAllGraceful() {
```

Comment:
A really fancy implementation could track what client certificate was used for the handshake and only close connections that match (this would work for exec since it knows both the old and the new cert).  I decided that was not worth the complexity.

```suggestion
func (c *ConnectionTracker) CloseAllGraceful(cert *tls.Certificate) {
```

Reply:
you can do this from the roundtripper using the httptrace package

https://pkg.go.dev/net/http/httptrace#WithClientTrace

you can get callbacks on the connection and tls state

Reply:
Using http trace requires controlling the context, which requires controlling the TLS dialer (you cannot do it from the round tripper because it can be skipped while still invoking the cert callback), at which point you can just directly instrument what you need.  Messing with the trace would require preserving any trace passed in by the caller, so it ends up being more code.  I also think it's less racey to mark the connection once the custom certificate is consumed instead of waiting for the handshake to complete.

Reply:
https://github.com/kubernetes/kubernetes/pull/112404/files#diff-60fb642abb39ae2e455ce58107134c510ae0c570045f5832446777980ae78176R1413-R1417 would fail if we relied on the round tripper.